File: dotnet/src/Vulgarity/PackReader.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Vulgarity
{
    /// <summary>Reads a masked pack document into terms and allowlist words.</summary>
    /// <remarks>
    /// A pack carries the same data as a seed document, in a form that holds no
    /// readable text. The bundled term lists ship as packs, so an assembly no
    /// longer prints the whole list under a tool such as strings.
    ///
    /// This is obfuscation, not encryption. The key sits in this file, so anyone
    /// who wants the list can still get it. The goal is only that nobody reads
    /// it by accident. tool/packlib.py holds the format and the encoder.
    /// </remarks>
    internal static class PackReader
    {
        public const int SupportedSchema = 1;

        private const int FlagHasAllow = 0x01;

        private static readonly byte[] Magic = { (byte)'V', (byte)'P', (byte)'K', (byte)'1' };

        // The mask key. It must match KEY in tool/packlib.py.
        private static readonly byte[] Key = Encoding.ASCII.GetBytes("vulgarity-pack-v1");

        // Encoding.UTF8 replaces a bad byte with U+FFFD and carries on. Dart's
        // utf8.decode throws. This throws too, so both ports refuse the same
        // corrupt pack instead of one of them loading mangled terms.
        private static readonly UTF8Encoding StrictUtf8 = new UTF8Encoding(false, true);

        /// <summary>True when these bytes start with the pack magic.</summary>
        public static bool Looks(byte[] pack)
        {
            if (pack == null || pack.Length < Magic.Length)
            {
                return false;
            }

            for (int i = 0; i < Magic.Length; i++)
            {
                if (pack[i] != Magic[i])
                {
                    return false;
                }
            }

            return true;
        }

        public static void Load(
            byte[] pack,
            string expectedProfile,
            List<VulgarityTerm> terms,
            List<string> allow)
        {
            if (pack == null)
            {
                throw new ArgumentNullException("pack");
            }

            if (!Looks(pack))
            {
                throw new FormatException("This is not a pack. The magic does not match.");
            }

            byte[] data = Unmask(pack, Magic.Length);
            if (data.Length < 2)
            {
                throw new FormatException("A pack must hold a header.");
            }

            int schema = data[0];
            if (schema != SupportedSchema)
            {
                throw new FormatException(
                    "This build reads pack schema " + SupportedSchema +
                    ". The file states schema " + schema + ".");
            }

            int flags = data[1];
            int pos = 2;

            // The profile pins the fold table the pack was built with. A stale file
            // then fails here instead of matching silently wrong.
            string profile = ReadText(data, ref pos);
            if (profile != expectedProfile)
            {
                throw new FormatException(
                    "This build implements fold profile '" + expectedProfile +
                    "'. The pack states '" + profile + "'.");
            }

            // The count is read from the pack, so it is not to be trusted with an
            // allocation. Grow the list while reading instead: a crafted count runs
            // the data out and fails as a malformed pack, not as an
            // OutOfMemoryException.
            int categoryCount = ReadVarint(data, ref pos);
            List<string> categories = new List<string>();
            for (int i = 0; i < categoryCount; i++)
            {
                categories.Add(ReadText(data, ref pos));
            }

            int entryCount = ReadVarint(data, ref pos);

            // An entry costs at least three bytes: a length, one character, and
            // the flags. Check that before building a list, so a hostile count
            // cannot make this allocate.
            if (entryCount > (data.Length - pos) / 3)
            {
                throw new FormatException("The pack claims more entries than it holds.");
            }

            for (int i = 0; i < entryCount; i++)
            {
                string term = ReadText(data, ref pos);
                if (term.Length == 0)
                {
                    throw new FormatException("A pack term must not be empty.");
                }

                if (pos >= data.Length)
                {
                    throw new FormatException("The pack ends inside an entry.");
                }

                int packed = data[pos];
                pos++;

                int index = packed >> 4;
                if (index >= categories.Count)
                {
                    throw new FormatException("A pack entry names a category the table does not hold.");
                }

                int severity = packed & 0x07;
                if (severity < 1 || severity > 5)
                {
                    throw new FormatException(
                        "Severity must be 1 to 5. Term '" + term + "' states " + severity + ".");
                }

                terms.Add(new VulgarityTerm(term, categories[index], severity, (packed & 0x08) != 0));
            }

            if ((flags & FlagHasAllow) != 0)
            {
                int allowCount = ReadVarint(data, ref pos);
                for (int i = 0; i < allowCount; i++)
                {
                    string word = ReadText(data, ref pos);
                    if (word.Length != 0)
                    {
                        allow.Add(word);
                    }
                }
            }

            // Nothing may follow. This catches a truncated pack and a padded one
            // in the same line.
            if (pos != data.Length)
            {
                throw new FormatException("The pack holds bytes after its last entry.");
            }
        }

        /// <summary>Strips the magic and XORs the rest with the keystream.</summary>
        private static byte[] Unmask(byte[] pack, int offset)
        {
            int length = pack.Length - offset;
            byte[] box = new byte[256];
            for (int i = 0; i < 256; i++)
            {
                box[i] = (byte)i;
            }

            int j = 0;
            for (int i = 0; i < 256; i++)
            {
                j = (j + box[i] + Key[i % Key.Length]) & 0xFF;
                byte swap = box[i];
                box[i] = box[j];
                box[j] = swap;
            }

            byte[] out_ = new byte[length];
            int a = 0;
            int b = 0;
            for (int i = 0; i < length; i++)
            {
                a = (a + 1) & 0xFF;
                b = (b + box[a]) & 0xFF;
                byte swap = box[a];
                box[a] = box[b];
                box[b] = swap;
                out_[i] = (byte)(pack[offset + i] ^ box[(box[a] + box[b]) & 0xFF]);
            }

            return out_;
        }

        /// <summary>Reads an unsigned LEB128 number, capped at 31 bits.</summary>
        /// <remarks>
        /// <para>
        /// The cap is what keeps every runtime reading the same pack. Dart compiles
        /// its shift to JavaScript's on the web, which truncates to 32 bits, so a
        /// value of 2^32+5 would read as 5 there and as 2^32+5 on the VM. Rather
        /// than let a crafted pack load on one and fail on the other, a number wider
        /// than 31 bits is refused everywhere. It also keeps the result inside an
        /// int here, where a wider one used to overflow.
        /// </para>
        /// <para>
        /// In practice that means a fifth byte carries three bits at most, and there
        /// is never a sixth. tool/packlib.py refuses to write what this refuses to
        /// read.
        /// </para>
        /// </remarks>
        private static int ReadVarint(byte[] data, ref int pos)
        {
            int value = 0;
            int shift = 0;
            while (true)
            {
                if (pos >= data.Length)
                {
                    throw new FormatException("The pack ends inside a number.");
                }

                int b = data[pos];
                pos++;

                // At the fifth byte only the low three bits are left, and a
                // continuation bit would ask for a sixth. Either is over the cap.
                if (shift == 28 && b > 0x07)
                {
                    throw new FormatException("A pack number runs too long.");
                }

                value |= (b & 0x7F) << shift;
                if (b < 0x80)
                {
                    return value;
                }

                shift += 7;
                if (shift > 28)
                {
                    throw new FormatException("A pack number runs too long.");
                }
            }
        }

        private static string ReadText(byte[] data, ref int pos)
        {
            // The length is capped at 31 bits, so it is never negative. The
            // subtraction, rather than pos + length, keeps the check itself from
            // overflowing on a length near int.MaxValue.
            int length = ReadVarint(data, ref pos);
            if (length < 0 || length > data.Length - pos)
            {
                throw new FormatException("The pack ends inside a string.");
            }

            string text;
            try
            {
                text = StrictUtf8.GetString(data, pos, length);
            }
            catch (DecoderFallbackException)
            {
                throw new FormatException("The pack holds text that is not valid UTF-8.");
            }

            pos += length;
            return text;
        }
    }
}

File: dart/lib/src/pack_reader.dart
import 'dart:convert';
import 'dart:typed_data';

import 'model/vulgarity_term.dart';

/// The pack schema this build reads.
const int kSupportedPackSchema = 1;

/// The magic every pack starts with. It is never masked, so a reader can
/// recognise the format without the key.
const List<int> kPackMagic = <int>[0x56, 0x50, 0x4B, 0x31]; // "VPK1"

const int _flagHasAllow = 0x01;

// The mask key. It must match KEY in tool/packlib.py.
const List<int> _key = <int>[
  0x76, 0x75, 0x6C, 0x67, 0x61, 0x72, 0x69, 0x74, 0x79, // "vulgarity"
  0x2D, 0x70, 0x61, 0x63, 0x6B, // "-pack"
  0x2D, 0x76, 0x31, // "-v1"
];

/// True when these bytes start with the pack magic.
bool looksLikePack(List<int> pack) {
  if (pack.length < kPackMagic.length) {
    return false;
  }
  for (int i = 0; i < kPackMagic.length; i++) {
    if (pack[i] != kPackMagic[i]) {
      return false;
    }
  }
  return true;
}

/// Reads a masked pack document into terms and allowlist words.
///
/// A pack carries the same data as a seed document, in a form that holds no
/// readable text. The bundled term lists ship as packs, so a compiled app no
/// longer carries the whole list as readable strings.
///
/// This is obfuscation, not encryption. The key sits in this file, so anyone
/// who wants the list can still get it. The goal is only that nobody reads it
/// by accident. `tool/packlib.py` holds the format and the encoder.
void loadPack(
  List<int> pack,
  String expectedProfile,
  List<VulgarityTerm> terms,
  List<String> allow,
) {
  if (!looksLikePack(pack)) {
    throw const FormatException(
        'This is not a pack. The magic does not match.');
  }

  final Uint8List data = _unmask(pack, kPackMagic.length);
  if (data.length < 2) {
    throw const FormatException('A pack must hold a header.');
  }

  final int schema = data[0];
  if (schema != kSupportedPackSchema) {
    throw FormatException('This build reads pack schema $kSupportedPackSchema. '
        'The file states schema $schema.');
  }

  final int flags = data[1];
  final _Cursor at = _Cursor(2);

  // The profile pins the fold table the pack was built with. A stale file then
  // fails here instead of matching silently wrong.
  final String profile = _readText(data, at);
  if (profile != expectedProfile) {
    throw FormatException(
        "This build implements fold profile '$expectedProfile'. "
        "The pack states '$profile'.");
  }

  final int categoryCount = _readVarint(data, at);
  final List<String> categories = <String>[
    for (int i = 0; i < categoryCount; i++) _readText(data, at),
  ];

  final int entryCount = _readVarint(data, at);

  // An entry costs at least three bytes: a length, one character, and the
  // flags. Check that before building a list, so a hostile count cannot make
  // this allocate.
  if (entryCount > (data.length - at.pos) ~/ 3) {
    throw const FormatException('The pack claims more entries than it holds.');
  }

  for (int i = 0; i < entryCount; i++) {
    final String term = _readText(data, at);
    if (term.isEmpty) {
      throw const FormatException('A pack term must not be empty.');
    }

    if (at.pos >= data.length) {
      throw const FormatException('The pack ends inside an entry.');
    }

    final int packed = data[at.pos];
    at.pos++;

    final int index = packed >> 4;
    if (index >= categoryCount) {
      throw const FormatException(
          'A pack entry names a category the table does not hold.');
    }

    final int severity = packed & 0x07;
    if (severity < 1 || severity > 5) {
      throw FormatException(
          "Severity must be 1 to 5. Term '$term' states $severity.");
    }

    terms.add(VulgarityTerm(
      term,
      categories[index],
      severity,
      (packed & 0x08) != 0,
    ));
  }

  if (flags & _flagHasAllow != 0) {
    final int allowCount = _readVarint(data, at);
    for (int i = 0; i < allowCount; i++) {
      final String word = _readText(data, at);
      if (word.isNotEmpty) {
        allow.add(word);
      }
    }
  }

  // Nothing may follow. This catches a truncated pack and a padded one in the
  // same line.
  if (at.pos != data.length) {
    throw const FormatException('The pack holds bytes after its last entry.');
  }
}

/// Strips the magic and XORs the rest with the keystream.
///
/// Every value here is one byte, and every step is `+` and `& 0xFF`. That keeps
/// the result identical on the Dart VM, on dart2js and on dart2wasm, where an
/// `int` is a JavaScript double and wider arithmetic would not agree.
Uint8List _unmask(List<int> pack, int offset) {
  final Uint8List box = Uint8List(256);
  for (int i = 0; i < 256; i++) {
    box[i] = i;
  }

  int j = 0;
  for (int i = 0; i < 256; i++) {
    j = (j + box[i] + _key[i % _key.length]) & 0xFF;
    final int swap = box[i];
    box[i] = box[j];
    box[j] = swap;
  }

  final int length = pack.length - offset;
  final Uint8List out = Uint8List(length);
  int a = 0;
  int b = 0;
  for (int i = 0; i < length; i++) {
    a = (a + 1) & 0xFF;
    b = (b + box[a]) & 0xFF;
    final int swap = box[a];
    box[a] = box[b];
    box[b] = swap;
    out[i] = pack[offset + i] ^ box[(box[a] + box[b]) & 0xFF];
  }

  return out;
}

class _Cursor {
  _Cursor(this.pos);

  int pos;
}

int _readVarint(Uint8List data, _Cursor at) {
  int value = 0;
  int shift = 0;
  while (true) {
    if (at.pos >= data.length) {
      throw const FormatException('The pack ends inside a number.');
    }

    final int b = data[at.pos];
    at.pos++;
    value |= (b & 0x7F) << shift;
    if (b < 0x80) {
      return value;
    }

    shift += 7;
    if (shift > 28) {
      throw const FormatException('A pack number runs too long.');
    }
  }
}

String _readText(Uint8List data, _Cursor at) {
  final int length = _readVarint(data, at);
  if (at.pos + length > data.length) {
    throw const FormatException('The pack ends inside a string.');
  }
  final String text = utf8.decode(data.sublist(at.pos, at.pos + length));
  at.pos += length;
  return text;
}
